### PR TITLE
Fixes for three possibles leaks

### PR DESCRIPTION
**Pull request by FdaSilvaYY**

### CODE REVIEW

**Comment by richsalz** [comment]
I don't see how the chan ges in tasn_dec are needed.

File: crypto/engine/eng_dyn.c [hunk 1]
```diff
@@ -231,6 +231,8 @@ static int dynamic_set_data_ctx(ENGINE *e, dynamic_data_ctx **ctx)
      * If we lost the race to set the context, c is non-NULL and *ctx is the
      * context of the thread that won.
      */
+    if (c)
+        sk_OPENSSL_STRING_free(c->dirs);
```
Comment (kroeckx) [inline crypto/engine/eng_dyn.c]: The check for if (c) is unneeded.

**Comment by kroeckx** [comment]
@richsalz: asn1_collect_data() can call collect_data() which calls BUF_MEM_grow_clean().  So the BUF_MEM_grow_clean() in asn1_d2i_ex_primitive() that just returned 0 could leak.

**Comment by richsalz** [comment]
Ah, thanks!

File: crypto/engine/eng_dyn.c (hunk 1 above)
Comment (FdaSilvaYY) [inline crypto/engine/eng_dyn.c]: A few lines of code before, there is a c = NULL; isn't it ?

Comment (kroeckx) [inline crypto/engine/eng_dyn.c]: You're right.

**Comment by kroeckx** [comment]
You seem to have added patches from master.  Can you rebase this instead and force an update?

**Comment by FdaSilvaYY** [comment]
Yes, I'm trying ;)

**Comment by kroeckx** [comment]
+1

**Pull request closed without merging**